src/System.Text.Encodings.Web.Utf8/BufferEncoder.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+
+namespace System.Text.Encodings.Web.Buffers
+{
+    public abstract class BufferEncoder : IBufferOperation
+    {
+        public abstract OperationStatus Encode(ReadOnlySpan<byte> input, Span<byte> output, out int consumed, out int written);
+        public virtual bool TryEncode(ReadOnlySpan<byte> input, Span<byte> output, out int written)
+        {
+            if (Encode(input, output, out int consumed, out written) == OperationStatus.Done)
+            {
+                return true;
+            }
+            return false;
+        }
+        public virtual OperationStatus EncodeInPlace(Span<byte> buffer, int length, out int written)
```

Comment:
Why is this a virtual method instead of an abstract one like Encode?

Reply:
I think many encoders won't be able to implement it, i.e. they will return not supported.